### DISCUSSION TITLE
Peers with lower bandwidth start more slowly as well.

### DISCUSSION THREAD

**Discussion by whr819987540** [anacrolix/torrent#899] [Ideas]
I have a cluster where there is one seeder and one hundred leeachers. The leechers have different network bandwidth. They start downloading at the same time. For one of the slower leeachers, the dowload log is shown as below.

```bash
3.006256614s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
6.007747097s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
9.008679463s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
12.009806195s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
15.010617602s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
18.01109505s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
21.012041165s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
24.012589419s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
27.013742126s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
30.014783721s: downloading "1.mp4": 0 B/171 MB(0 Bytes written to memory), 0/1302 pieces completed (0 partial): 0 B/s
33.015707056s: downloading "1.mp4": 1.1 MB/171 MB(0 Bytes written to memory), 7/1302 pieces completed (4 partial): 366 kB/s
36.016123573s: downloading "1.mp4": 2.9 MB/171 MB(0 Bytes written to memory), 16/1302 pieces completed (14 partial): 601 kB/s
39.017902407s: downloading "1.mp4": 8.3 MB/171 MB(0 Bytes written to memory), 48/1302 pieces completed (37 partial): 1.8 MB/s
42.019238495s: downloading "1.mp4": 17 MB/171 MB(0 Bytes written to memory), 106/1302 pieces completed (60 partial): 3.0 MB/s
45.020565591s: downloading "1.mp4": 26 MB/171 MB(0 Bytes written to memory), 176/1302 pieces completed (56 partial): 2.8 MB/s
48.021896843s: downloading "1.mp4": 33 MB/171 MB(0 Bytes written to memory), 214/1302 pieces completed (78 partial): 2.4 MB/s
51.023031154s: downloading "1.mp4": 41 MB/171 MB(0 Bytes written to memory), 264/1302 pieces completed (88 partial): 2.5 MB/s
54.024720515s: downloading "1.mp4": 48 MB/171 MB(0 Bytes written to memory), 308/1302 pieces completed (115 partial): 2.4 MB/s
57.025572858s: downloading "1.mp4": 55 MB/171 MB(0 Bytes written to memory), 341/1302 pieces completed (158 partial): 2.4 MB/s
```

The slow leecher is idle at the beginning. I am sure that it has found the seeder and other leechers as I have set the announce interval to 1s after I have found this problem. However, even when the annouce interval is set to 1s, this problem still exists.

Can you give me some ideas to debug?

**Comment by anacrolix** [comment]
It's most likely that all 100 leechers are not able to connect at the same time to the seeder. The current behaviour is currently to favour peers that have a certain peer ID, then to favour based on whether the peers want to leech or seed depending on the state of the local client. So in the case of the seeding client, it will prefer peers with a particular peer ID and then kick the ones that aren't showing interest in leeching from it.

I guess you would go about debugging this by seeing if the leechers that take a while to start downloading are actually connecting to the seeder, and how long they are connected to it before they start to receive data.

If there is a bug here, it could be that leechers will are not kicking or favouring peers that don't have data that they want. I can't say for certain that isn't the case without digging in to double check. Let me know what you think.

**Comment by whr819987540** [comment]
You are right. The number of max connections that could be established is limited(50, however I have 100 leechers. So not every leecher could connect to the server. This is also proved as I count the chunks receiving from the initial seeder.). I count the total download time of all leechers, which decreases from 81.4 to 74.6 with the number of max connections increasing to 200.

**Reply by whr819987540** [reply]
It can be concluded that it matters to adjust the config under different circumstances.

**Reply by anacrolix** [reply]
Interesting. Thanks for confirming. You are welcome to make a PR to document this kind of thing!
